components/bin/build

```diff
@@ -0,0 +1,318 @@
+ *
+ * @param {string} base    The root directory for the .ts files
+ * @param {string} name    The path of the file relative to the base
+ */
+function processFile(base, name) {
+    console.log(name);
+    const file = fs.readFileSync(path.resolve(base, name)).toString();
+    const parts =
+        file.split(/(^export(?:\s+default)?(?:\s+abstract)?\s+[^ {*}]+\s+[a-zA-Z0-9_.$]+)/m);
+    const objects = processParts(parts);
+    const lines = processLines(name, objects);
+    makeFile(name, lines);
+    if (objects.length) {
+        PACKAGE.push(name);
+    }
+}
+
+/**
+ * Process the export lines to record the ones that need to be added to the
+ *   library file.
+ *
+ * @param {string[]} parts       The file broken into export lines and the text between them
+ * @return {string[]}            An array of names of exported objects
+ */
+function processParts(parts) {
+    const objects = [];
+    for (let i = 1; i < parts.length; i += 2) {
+        const words = parts[i].split(/\s+/);
+        const type = words[words.length - 2];
+        const name = words[words.length - 1];
+
+        if (words[1] === 'default') {
+
+            objects.push('default');
+
+        } else {
+
+            switch (type) {
+              case 'type':
+              case 'interface':
+                break;
+                
+              case 'let':
+              case 'const':
+              case 'var':
+              case 'function':
+              case 'class':
+              case 'namespace':
+                objects.push(name);
+                break;
+                
+              case 'default':
+                objects.push('default');
+                break;
+                
+              default:
+                console.log("   Can't process '" + parts[i] + "'");
+                break;
+            }
+        }
+    }
+    return objects;
+}
+
+/**
+ * Create the lines of the lib file using the object names from the file.
+ *
+ * @param {string} name       The path of the file relative to the root .ts directory
+ * @param {string[]} objects  Array of names of the object exported by the file
+ * @return {string[]}         Array of lines for the contents of the library file
+ */
+function processLines(file, objects) {
+    if (objects.length === 0) return [];
+    const dir = path.dirname(file).replace(/^\.$/, '');
+    const dots = dir.replace(/[^\/]+/g, '..') || '.';
+    const relative = path.join(dots, '..', MATHJAX3, dir, path.basename(file)).replace(/\.ts$/, '.js');
+    const name = path.parse(file).name;
+    const lines = [
+        '"use strict";',
+        `Object.defineProperty(exports, '__esModule', {value: true});`
+    ];
+    let source = (dir.replace(/\//g, '.') + '.' + name).replace(/^\./, '')
+        .replace(/\.[^.]*/g, (x) => (x.substr(1).match(/[^a-zA-Z_]/) ? '[\'' + x.substr(1) + '\']' : x));
+    if (exists(path.resolve(MATHJAX3, file.replace(/\.ts$/, '')))) source += '_ts';
```

Comment:
Missing braces.

Why do we need a  `_ts`  on the filename?

Reply:
As we discussed face-to-face, since there can be directories with the same base name as a `.ts` file, there needs to be a way to distinguish the two.  For example, there are both `output/svg.ts` and `output/svg` (which contains things like `output/svg/Wrapper.ts`).  Since `output/svg/Wrapper.ts` will be using `MathJax._.output.svg.Wrapper = ...`, we use `MathJax._.output.svg_ts = ...` for the contents of the `output/svg.ts` file.  It is unpleasant, but necessary.